src/main/java/uk/gov/justice/digital/writer/Writer.java

```diff
@@ -61,70 +58,16 @@ protected static void writeCdcRecords(
     ) throws DataStorageRetriesExhaustedException {
         logger.info("Applying {} CDC records to deltalake table: {}", validRecords.count(), destinationPath);
 
-        for (Row row: validRecords.collectAsList()) {
-            processRow(spark, storage, destinationPath, primaryKey, row);
-        }
+        val inserts = validRecords.filter(col(OPERATION).equalTo(Insert.getName())).orderBy(TIMESTAMP);
+        val updates = validRecords.filter(col(OPERATION).equalTo(Update.getName())).orderBy(TIMESTAMP);
+        val deletes = validRecords.filter(col(OPERATION).equalTo(Delete.getName())).orderBy(TIMESTAMP);
+
+        storage.upsertRecords(spark, destinationPath, inserts.drop(OPERATION, TIMESTAMP), primaryKey);
+        storage.updateRecords(spark, destinationPath, updates.drop(OPERATION, TIMESTAMP), primaryKey);
```

Comment:
What happens when there are multiple updates for the same PK? Does it apply them all in timestamp order?

Reply:
Good question. I think Delta has a feature for this. Taking a look.

Reply:
If not and if every update contains the full up to date record we could probably drop all but the latest row for a given PK using a window function or something.

Reply:
This is now implemented. Before writing, the latest records are selected when there are more than one record for a primary key.